### ISSUE TITLE
[mongodb_atlas]: Parsing errors in multiple datasets/datastreams and agent instability in Beta version

### ISSUE THREAD

**Issue by rheppe** [elastic/integrations#11576]
### Integration Name

MongoDB Atlas [mongodb_atlas]

### Dataset Name

mongodb_atlas-alert,mongodb_atlas-hardware,mongodb_atlas-mongod_database,mongodb_atlas-organization,mongodb_atlas-process,mongodb_atlas-project

### Integration Version

0.0.8

### Agent Version

8.15.1

### Agent Output Type

elasticsearch

### Elasticsearch Version

8.15.0

### OS Version and Architecture

Elastic Docker Image hosted on Azure Container Apps (x86_64/linux)

### Error Message

Please see the error-log.json 

### Event Original

Please see the request-trace.json

### What did you do?

I did some digging and isolated the error to a particular request. In the attached 'error-log.json' you can see the error being logged. The 'request-trace.json' file contains the corresponding request to the MongoDB API.

If I understand the error logs correctly, the error occurs when parsing the response from MongoDB in line 384 of the file input.go:

```
`"log.origin": {
"file.line": 384,
"[file.name](https://file.name/)": "cel/input.go",
"function": "[github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1](https://github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1)"
},`
```

When browsing Elastic GitHub I can fin the line here: https://github.com/elastic/beats/blob/main/x-pack/filebeat/input/cel/input.go#L384:

```
`if err, ok := e["error"]; ok {
env.UpdateStatus(status.Degraded, fmt.Sprintf("single event error object returned by evaluation: %s", mapstr.M{"error": err}))
} else {
env.UpdateStatus(status.Degraded, "single event object returned by evaluation")
}`
```

I don't have any knowledge of GO, so I can't determine what exactly is going wrong. But I think your developers should be able to quickly find out what the issue is with this information.

### What did you see?

I did some digging and isolated the error to a particular request. In the attached 'error-log.json' you can see the error being logged. The 'request-trace.json' file contains the corresponding request to the MongoDB API.

If I understand the error logs correctly, the error occurs when parsing the response from MongoDB in line 384 of the file input.go:

```
`"log.origin": {
"file.line": 384,
"[file.name](https://file.name/)": "cel/input.go",
"function": "[github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1](https://github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1)"
},`
```

When browsing Elastic GitHub I can fin the line here: https://github.com/elastic/beats/blob/main/x-pack/filebeat/input/cel/input.go#L384:

```
`if err, ok := e["error"]; ok {
env.UpdateStatus(status.Degraded, fmt.Sprintf("single event error object returned by evaluation: %s", mapstr.M{"error": err}))
} else {
env.UpdateStatus(status.Degraded, "single event object returned by evaluation")
}`
```

I don't have any knowledge of GO, so I can't determine what exactly is going wrong. But I think your developers should be able to quickly find out what the issue is with this information.

### What did you expect to see?

Parsed output

**Comment by rheppe** [comment]
Additional information, from customer.
```
I did some digging and isolated the error to a particular request. In the attached 'error-log.json' you can see the error being logged. The 'request-trace.json' file contains the corresponding request to the MongoDB API.

If I understand the error logs correctly, the error occurs when parsing the response from MongoDB in line 384 of the file input.go:

`"log.origin": {
"file.line": 384,
"[file.name](https://file.name/)": "cel/input.go",
"function": "[github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1](https://github.com/elastic/beats/v7/x-pack/filebeat/input/cel.input.run.func1)"
},`

When browsing Elastic GitHub I can fin the line here: https://github.com/elastic/beats/blob/main/x-pack/filebeat/input/cel/input.go#L384:

`if err, ok := e["error"]; ok {
env.UpdateStatus(status.Degraded, fmt.Sprintf("single event error object returned by evaluation: %s", mapstr.M{"error": err}))
} else {
env.UpdateStatus(status.Degraded, "single event object returned by evaluation")
}`

I don't have any knowledge of GO, so I can't determine what exactly is going wrong. But I think your developers should be able to quickly find out what the issue is with this information.
```

```
Here are the missing items:
* Datasets we are using:
* MongoDB Atlas Alert logs (mongodb_atlas-alert)
* MongoDB Atlas Hardware metrics (mongodb_atlas-hardware)
* MongoDB Atlas Mongod Database logs (mongodb_atlas-mongod_database)
* MongoDB Atlas Organization logs (mongodb_atlas-organization)
* MongoDB Atlas Process metrics (mongodb_atlas-process)
* MongoDB Atlas Project logs (mongodb_atlas-project)
* OS version and Architecture:
* We are using the **elastic-agent/elastic-agent:8.15.1** Elastic Docker Image hosted on Azure Container Apps (x86_64/linux)
* Agent Output type:
* I assume Elasticsearch since we are on Elastic Cloud and AFAIK we are not using Logstash.

I think it is highly unlikely the issue is caused by the Agent version, since the error is pretty clear that it's coming from the integration code. However, I will deploy an v8.15.0 version of the agent to make sure the version is aligned with the Kibana/Elastic version.
```

Files:
```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: request-trace.json
You can download the file here: https://upload.elastic.co/d/73fcc15b88a6856348d0ea91898ff1b108785a983dc45d168edfb5dbe045d87e.
Use this Authorization Token: 3541c0c08146bb0e
```

```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: error-log.json
You can download the file here: https://upload.elastic.co/d/cd52d4d0ece13354286ce225c37f996741484801eead5337f4ac2599272a9a31.
Use this Authorization Token: 27156e312b0f1761
```

```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: elastic-agent-diagnostics-2024-10-28T12-53-02Z-00.zip
You can download the file here: https://upload.elastic.co/d/258d6ff129e7109693bd97930bc7129529da8385b57a1d674b8f62b60108fc74.
Use this Authorization Token: 832e771c708dd784
```

```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: elastic-agent-unhealthy.png
You can download the file here: https://upload.elastic.co/d/bcf06eb61aab1f9910437c70f90d4a111cc660ecfc343a330dc1bfe466d99825.
Use this Authorization Token: 385c0f19011fe257
```

```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: http-request-trace-cel-mongodb_atlas.process-bd4275c7-86aa-4437-ab22-6a25...json
You can download the file here: https://upload.elastic.co/d/d59df5cefe7b3200faebec87e2c46474240af6ed3cecadfa9c23a51ea2b31ee8.
Use this Authorization Token: b9a8b0d6c71abab9
```

```
Upload for Elastic Cloud - [p_jcorbett@ws.wsfgrp.net](mailto:p_jcorbett@ws.wsfgrp.net) - 01769025 is complete.
File name: elastic-agent-20241017-4.ndjson.json
You can download the file here: https://upload.elastic.co/d/b964230fa7bd65d7c8f00542da93e0ab2592818417266e8296aa616bf9058105.
Use this Authorization Token: d08c5f01213fc588
```

**Comment by efd6** [comment]
@rheppe The code that you are linking to is working as intended. That it is triggering the logging that you see is telling you that an error condition happened in attempts to collect the data from the API. The message in the log line would indicate what that error was. I've taken a look at the relevant code in the integration and I see that the way the collectors are written, fewer of these errors are possible than should be possible according to the input's design. This should probably be improved.

Providing the message (and only the message) of the logged errors would be helpful for the integration maintainers to be able to address the issue that you have raised here.

**Comment by jsuurd** [comment]
@efd6 The issue is that Elastic Agent is in degraded state because the MongoDB Atlas integration is raising the following error when parsing the payload returned by the MongoDB API:
```
    "log.level": "error",
    "@timestamp": "2024-10-28T12:45:22.879Z",
    "message": "single event object returned by evaluation",
```

The [error-log.json](https://github.com/user-attachments/files/17601100/error-log.json) and [request-trace.json](https://github.com/user-attachments/files/17601107/request-trace.json) that caused the error was provided to Elastic Support and they are mentioned in the issue description. I uploaded them to this messages as well.

I also provided the full diagnostic zip, @rheppe should be able to provide this to you.

Thanks,
Jeroen

**Comment by efd6** [comment]
Thanks @jsuurd. Yes, I agree, that is not correct use of the input.

@harnish-elastic The places where this behaviour is being specified are here:
- [disk](https://github.com/elastic/integrations/blob/31a158dca4f6632687638cafca69526043f83e50/packages/mongodb_atlas/data_stream/disk/agent/stream/input.yml.hbs#L126-L135)
- [hardware](https://github.com/elastic/integrations/blob/31a158dca4f6632687638cafca69526043f83e50/packages/mongodb_atlas/data_stream/hardware/agent/stream/input.yml.hbs#L57-L63)
- [process](https://github.com/elastic/integrations/blob/31a158dca4f6632687638cafca69526043f83e50/packages/mongodb_atlas/data_stream/process/agent/stream/input.yml.hbs#L57-L61)

AFAICS these all return a single object for the `events` field. This has the effect of raising an error condition. This does not appear to be the intention in those locations. This is the converse of the situation for the error cases that we discussed off-line.

**Comment by jsuurd** [comment]
@harnish-elastic Is there any ETA when the fix of this issue will be available?

**Comment by harnish-elastic** [comment]
> [@harnish-elastic](https://github.com/harnish-elastic) Is there any ETA when the fix of this issue will be available?

[PR](https://github.com/elastic/integrations/pull/11650) has already been raised. Once I get approval on the PR, the changes will be available on mongodb_atlas version `0.0.10`
